File: src/main/java/io/scriptor/machine/FPRFile.java
package io.scriptor.machine;

import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;

public interface FPRFile {

    void reset();

    void dump(final @NotNull PrintStream out);

    float getf(final int reg);

    double getd(final int reg);

    void putf(final int reg, final float val);

    void putd(final int reg, final double val);

    int getfr(final int reg);

    long getdr(final int reg);

    void putfr(final int reg, final int val);

    void putdr(final int reg, final long val);
}

File: src/main/java/io/scriptor/machine/FPRFile.kt
package io.scriptor.machine

interface FPRFile : Device {

    fun getf(reg: UInt): Float = Float.fromBits(getfr(reg).toInt())
    fun getd(reg: UInt): Double = Double.fromBits(getdr(reg).toLong())

    fun put(reg: UInt, value: Float) = put(reg, value.toRawBits().toUInt())
    fun put(reg: UInt, value: Double) = put(reg, value.toRawBits().toULong())

    fun getfr(reg: UInt): UInt
    fun getdr(reg: UInt): ULong

    fun put(reg: UInt, value: UInt)
    fun put(reg: UInt, value: ULong)

    operator fun set(reg: UInt, value: Float) = put(reg, value)
    operator fun set(reg: UInt, value: Double) = put(reg, value)
    operator fun set(reg: UInt, value: UInt) = put(reg, value)
    operator fun set(reg: UInt, value: ULong) = put(reg, value)
}
